kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogReplaySuite.scala

```diff
@@ -13,84 +13,264 @@
+
+    // Append data0
+    var data0_files: Array[File] = Array.empty
+    withGoldenTable("snapshot-data0") { tablePath =>
+      data0_files = getDirDataFiles(tablePath) // data0 files
+      verifySnapshot(tablePath, data0_files, 0)
+    }
+
+    // Append data1
+    var data0_data1_files: Array[File] = Array.empty
+    withGoldenTable("snapshot-data1") { tablePath =>
+      data0_data1_files = getDirDataFiles(tablePath) // data0 & data1 files
+      verifySnapshot(tablePath, data0_data1_files, 1)
+    }
+
+    // Overwrite with data2
+    var data2_files: Array[File] = Array.empty
+    withGoldenTable("snapshot-data2") { tablePath =>
+      // we have overwritten files for data0 & data1; only data2 files should remain
+      data2_files = getDirDataFiles(tablePath)
+        .filterNot(f => data0_data1_files.exists(_.getName == f.getName))
+      verifySnapshot(tablePath, data2_files, 2)
+    }
+
+    // Append data3
+    withGoldenTable("snapshot-data3") { tablePath =>
+      // we have overwritten files for data0 & data1; only data2 & data3 files should remain
+      val data2_data3_files = getDirDataFiles(tablePath)
+        .filterNot(f => data0_data1_files.exists(_.getName == f.getName))
+      verifySnapshot(tablePath, data2_data3_files, 3)
+    }
+
+    // Delete data2 files
+    withGoldenTable("snapshot-data2-deleted") { tablePath =>
+      // we have overwritten files for data0 & data1, and deleted data2 files; only data3 files
+      // should remain
+      val data3_files = getDirDataFiles(tablePath)
+        .filterNot(f => data0_data1_files.exists(_.getName == f.getName))
+        .filterNot(f => data2_files.exists(_.getName == f.getName))
+      verifySnapshot(tablePath, data3_files, 4)
+    }
+
+    // Repartition into 2 files
+    withGoldenTable("snapshot-repartitioned") { tablePath =>
+      val snapshot = latestSnapshot(tablePath)
+      assert(snapshot.getVersion(defaultTableClient) == 5)
+      val scanFileRows = collectScanFileRows(
+        snapshot.getScanBuilder(defaultTableClient).build())
+      assert(scanFileRows.length == 2)
+    }
+
+    // Vacuum
+    withGoldenTable("snapshot-vacuumed") { tablePath =>
+      // all remaining dir data files should be needed for current snapshot version
+      // vacuum doesn't change the snapshot version
+      verifySnapshot(tablePath, getDirDataFiles(tablePath), 5)
+    }
+  }
+
+  test("DV cases with same path different DV keys") {
+    val snapshot = latestSnapshot(goldenTablePath("log-replay-dv-key-cases"))
+    val scanFileRows = collectScanFileRows(
+      snapshot.getScanBuilder(defaultTableClient).build()
+    )
+    assert(scanFileRows.length == 1) // there should only be 1 add file
+    val dv = InternalScanFileUtils.getDeletionVectorDescriptorFromRow(scanFileRows.head)
+    assert(dv.getCardinality == 3) // dv cardinality should be 3
+  }
+
+  test("special characters in path") {
+    val snapshot = latestSnapshot(goldenTablePath("log-replay-special-characters"))
```

Comment:
If not exists, good to add a test that scan file rows returned contains the path correctly for an AddFile with special characters path.